detox/src/devices/drivers/android/AndroidDriver.js

```diff
@@ -256,56 +204,37 @@ class AndroidDriver extends DeviceDriverBase {
     return testApkPath;
   }
 
-  async _launchInstrumentationProcess(deviceId, bundleId, userLaunchArgs) {
-    const serverPort = new URL(this.client.configuration.server).port;
-    await this.adb.reverse(deviceId, serverPort);
-
-    this.instrumentationLogsParser = new InstrumentationLogsParser();
-    this.instrumentation.setTerminationFn(() => this._onInstrumentationTerminated(deviceId, serverPort));
-    this.instrumentation.setLogListenFn(this._extractStackTraceFromInstrumLogs.bind(this));
-    await this.instrumentation.launch(deviceId, bundleId, userLaunchArgs);
-  }
-
-  async _onInstrumentationTerminated(deviceId, serverPort) {
-    await this.adb.reverseRemove(deviceId, serverPort);
-    this.instrumentationCloseListener();
-
-    this.instrumentation.setTerminationFn(null);
-    this.instrumentation.setLogListenFn(null);
-  }
-
-  async _terminateInstrumentation() {
-    await this.instrumentation.terminate();
-    this.instrumentationCloseListener();
-
-    this.instrumentation.setTerminationFn(null);
-    this.instrumentation.setLogListenFn(null);
-  }
-
-  _extractStackTraceFromInstrumLogs(logsDump) {
-    this.instrumentationLogsParser.parse(logsDump);
-
-    if (this.instrumentationLogsParser.containsStackTraceLog(logsDump)) {
-      this.instrumentationStackTrace = this.instrumentationLogsParser.getStackTrace(logsDump);
+  async _modifyArgsForNotificationHandling(deviceId, bundleId, launchArgs) {
+    let _launchArgs = launchArgs;
+    if (launchArgs.detoxUserNotificationDataURL) {
+      const notificationPayloadTargetPath = await this._sendNotificationDataToDevice(launchArgs.detoxUserNotificationDataURL, deviceId);
+      _launchArgs = {
+        ...launchArgs,
+        detoxUserNotificationDataURL: notificationPayloadTargetPath,
+      }
     }
+    return _launchArgs;
   }
 
-  async _queryPID(deviceId, bundleId, waitAtStart = true) {
-    if (waitAtStart) {
+  async _launchApp(deviceId, bundleId, launchArgs) {
+    if (!this.instrumentation.isRunning()) {
+      await this._launchInstrumentationProcess(deviceId, bundleId, launchArgs);
       await sleep(500);
+    } else if (launchArgs.detoxURLOverride) {
+      await this._startActivityWithUrl(launchArgs.detoxURLOverride);
+    } else if (launchArgs.detoxUserNotificationDataURL) {
+      await this._startActivityFromNotification(launchArgs.detoxUserNotificationDataURL);
+    } else {
+      await this._resumeMainActivity();
     }
+  }
 
-    for (let attempts = 5; attempts > 0; attempts--) {
-      const pid = await this.adb.pidof(deviceId, bundleId);
-
-      if (pid > 0) {
-        return pid;
-      }
-
-      await sleep(1000);
-    }
```

Comment:
unified with the existing `retry()` util function though a new arg: `backoff: 'none'` (default is `linear`).